### ISSUE TITLE
SNOW-961482 High CPU usage in gosnowflake OCSP module

### ISSUE THREAD

**Issue by aldld** [snowflakedb/gosnowflake#942]
Please answer these questions before submitting your issue. 
In order to accurately debug the issue this information is required. Thanks!

1. What version of GO driver are you using?
1.6.24
   
2. What operating system and processor architecture are you using?
Linux amd64 (Debian 11 Bullseye)
   
3. What version of GO are you using?
go 1.20

4.Server version:* E.g. 1.90.1
You may get the server version by running a query:
```
SELECT CURRENT_VERSION();
```
5. What did you do?

In some recent load tests, we've observed a lot of CPU usage coming from the driver's [peer certificate verification routine](https://github.com/snowflakedb/gosnowflake/blob/master/ocsp.go#L789) during the TLS client handshake. From our profiling, most of the CPU usage comes from [(re-)parsing cached OCSP responses](https://github.com/snowflakedb/gosnowflake/blob/master/ocsp.go#L885). This entails verifying the cached response signature, which is a fairly expensive operation.

It seems like we re-verify the same cached OCSP responses many times because the cache entries are represented in memory as base64-encoded strings that must be parsed (and therefore verified) each time they are used. Is there any reason not to represent the cached OCSP responses in memory in their parsed form (ocsp.Response), and only verify a given signature once instead of on every handshake?

6. What did you expect to see?

I would expect the cached OCSP responses to only have to be parsed once, rather than re-parsing and re-validating the responses every time they are used.

7. Can you set logging to DEBUG and collect the logs?

   https://community.snowflake.com/s/article/How-to-generate-log-file-on-Snowflake-connectors
   
8. What is your Snowflake account identifier, if any? (Optional)

**Comment by sfc-gh-dszmolka** [comment]
hi and thank you for raising this issue with us ! before we delve deeper into the current implementation and how it could be optimalized (by the way, thank you for the great pointers!), just to be on the same page: which libraries/methods did you use to measure CPU consumption ?

just in case we would like to do the same before, and after a potential change, to be able to compare to the same baseline.
thank you in advance!

**Comment by aldld** [comment]
Hi @sfc-gh-dszmolka, thanks for the reply. To measure CPU consumption, we used Grafana Pyroscope to profile our application's CPU usage in a staging Kubernetes cluster. However, I imagine you'd be able to observe the same results using go's builtin benchmarking and pprof tools.

**Comment by tomas-karnagel** [comment]
We are seeing the same issue quite drastically even without load testing.
We switched to go1.20 and got severe delays in our query execution, which we first thought would be due to Snowflake issues but later found that the CPU consumption is through the roof and time is spend in the gosnowflake driver. We had to fall back to go1.19.

- gosnowflake version: 1.6.25
- SELECT CURRENT_VERSION();: 7.39.4
- go version: go 1.20
- Linux
